message/router.go

```diff
@@ -317,67 +329,108 @@ func (r *Router) Run(ctx context.Context) (err error) {
 		}
 	}
 
-	r.logger.Debug("Applying decorators", nil)
+	if err := r.RunHandlers(ctx); err != nil {
+		return err
+	}
+
+	close(r.running)
+
+	go r.closeWhenAllHandlersStopped()
+
+	<-r.closeCh
+	cancel()
+
+	r.logger.Info("Waiting for messages", watermill.LogFields{
+		"timeout": r.config.CloseTimeout,
+	})
+
+	<-r.closedCh
+
+	r.logger.Info("All messages processed", nil)
+
+	return nil
+}
+
+// RunHandlers runs all handlers that were added after Run().
+// RunHandlers is idempotent, so can be called multiple times safely.
+func (r *Router) RunHandlers(ctx context.Context) error {
+	r.handlersLock.Lock()
+	defer r.handlersLock.Unlock()
+
```

Comment:
Should we check if the router is running at this point?

Reply:
Good point :+1:. Fixed in a756811